### ISSUE TITLE
python functions added to trays do not trace usage statistics (Trac #2406)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2398]
see title

Migrated from https://code.icecube.wisc.edu/ticket/2406
```json
{
    "status": "closed", 
    "changetime": "2020-01-29T19:46:27", 
    "description": "see title", 
    "reporter": "chaack", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1580327187269558", 
    "component": "combo core", 
    "summary": "python functions added to trays do not trace usage statistics", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2020-01-29T18:28:40", 
    "milestone": "", 
    "owner": "nwhitehorn", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20200129T18:51:41: jvansanten commented:

> Some background: this happens because I3Module does its timing inside of Process(), but PythonFunction implements its own version of Process(). A possible solution would be do make PythonFunction implement Physics(), DAQ(), and all other named stop methods that just do the same thing as the current Process().

**Comment by icecube-trac** [comment]
Trac update at 20200129T18:53:44: nwhitehorn commented:

> Can't we just make PythonFunction's Process() do the same accounting as I3Module's version? We could also make a private Process_() that calls Process() and include the accounting there.

**Comment by icecube-trac** [comment]
Trac update at 20200129T18:58:19: jvansanten commented:

> Mostly because it involves either exposing more of I3Module's state to its daughter classes, or duplicating that state in the daughter class. Making PythonFunction behave as if it were a normal I3Module that happened to call a Python function in its Physics() method seems like a smaller change.

**Comment by icecube-trac** [comment]
Trac update at 20200129T19:01:26: nwhitehorn commented:

> The Process_() thing doesn't seem intrusive, it's just moving the ModuleTimer stuff a couple lines up to Process_() from Process() in I3Module.cxx. It's a 5-line diff (I'm happy to do it) that doesn't escape I3Module.cxx.

**Comment by icecube-trac** [comment]
Trac update at 20200129T19:06:56: jvansanten commented:

> Fine. Fire away then.

**Comment by icecube-trac** [comment]
Trac update at 20200129T19:46:27: 
* nwhitehorn changed status from "new" to "closed"
* nwhitehorn commented:

> In changeset 178701:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="178701"
> Always record time spent in Physics()/DAQ(), even if the module overrides
> Process(), as Python functions do.
> 
> This fixes #2398.
> ```


* nwhitehorn changed owner from "" to "nwhitehorn"
* nwhitehorn changed resolution from "" to "fixed"
